### PR TITLE
Introduce DeviceCredentials for verifying credentials.

### PR DESCRIPTION
**Pull request by sophokles73**
DeviceCredentials already provide validation logic specific to the
particular type of credentials they represent. The SecretValidator
concept is therefore no longer needed.

Signed-off-by: Kai Hudalla <kai.hudalla@bosch-si.com>

### CODE REVIEW

**Comment by sophokles73** [comment]
> Why? I thought it probably not always should be implemented via the java.security package, but maybe some crypto libs or any other to deal with the details of validation.
Or not have the password being stored in by Base64 encoding, or any other differences.
With the secret validators it would have been only to write another validator that feels responsible for this type and drop it into the classpath.
If I see it correctly, now the object itself must be somehow substituted since it carries the algorithm to validate in itself.

I see your point, however, I do not think that the current implementation allows for doing this. The `CredentialsUtils` class maintains a map of secret types to validators. It keeps one particular validator per type. We already have a validator for hashed passwords in our code base. So, if a user implements his own alternative version of a validator for hashed passwords and *drops it into the classpath* (to use your words ;-)) then which version will end up in the map and thus being used?
It is actually not that easy as it sounds to *drop it into the classpath*, this would at least require the user to build his own Docker image, right? I have my doubts that people will find it very desirable to (have to) implement their own secret validation logic. At least we tell others to **absolutely not** try to implement security related stuff themselves if there already is a working implementation based on standards.
So, if there is a need for an alternative implementation of the *hashed-password* type, I would be interested in a user opening an issue for it. We can then see what needs to be done (if anything needs to be done). The main issue I can see is that we do not support many secret types (yet). However, all of the ones we currently also envision (RawPublicKeys, PreSharedKeys) need to be implemented at a lower level (the TLS handshake) anyway. So, if we want to add a CoAP protocol adapter based on Californium, we would need to implement the particular interface provided by Cf for retrieving the secrets based on the plain Credentials API, i.e. the validation of the secrets would be done by Cf and not by a Hono SecretsValidator ...

File: adapters/mqtt-vertx/src/test/resources/logback-test.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    <appender-ref ref="STDOUT" />
+  </root>
+
+  <logger name="org.eclipse.hono.adapter" level="INFO"/>
+  <logger name="org.eclipse.hono.auth" level="INFO"/>
+  <logger name="org.eclipse.hono.config" level="INFO"/>
+  <logger name="org.eclipse.hono.connection" level="INFO"/>
+  <logger name="org.eclipse.hono.util" level="INFO"/>
+
+</configuration>
```
Comment (sysexcontrol) [inline adapters/mqtt-vertx/src/test/resources/logback-test.xml]: newline?

File: core/src/main/java/org/eclipse/hono/auth/DeviceCredentials.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * Verifies that the credentials provided by the device match the credentials
+     * that are on record for the device.
+     * 
+     * @param credentialsOnRecord The credentials for the device as returned by the
+     *                            <em>Credentials</em> API.
+     * @return {@code true} if the credentials provided by the device have been validated successfully.
+     * @throws IllegalArgumentException if the credentials on record do not contain any secrets.
+     */
+    boolean validate(CredentialsObject credentialsOnRecord);
+}
```
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/auth/DeviceCredentials.java]: newline

**Comment by sysexcontrol** [comment]
> I see your point, however, I do not think that the current implementation allows for doing this. The CredentialsUtils class maintains a map of secret types to validators. It keeps one particular validator per type. We already have a validator for hashed passwords in our code base. So, if a user implements his own alternative version of a validator for hashed passwords and drops it into the classpath (to use your words ;-)) then which version will end up in the map and thus being used?

Right - for now there was no support for having more than one validator per type.
That you just need to drop it into the classpath is not directly implemented in this version, I agree - it was thought like that but would need some effort to enable it. Build your own docker image would of course be necessary, but I do not consider that a big issue.

Anyway, I understand your concerns (especially about not encouring own security relevant implementations), and IMHO we can also handle this via issues.
So keep it simple and see if there are demands for other implementations - fine for me. Skip the validators mechanism then.

I would still like some decoupling of the username/password object from the algorithm, but not necessarily via the validator mechanism (maybe just via an interface). Would be cleaner to me, WDYT?

**Comment by sophokles73** [comment]
> I would still like some decoupling of the username/password object from the algorithm, but not necessarily via the validator mechanism (maybe just via an interface). Would be cleaner to me, WDYT?

Is this about being able to re-use the validation procedure in other `DeviceCredentials` implementations?
If so, then we could encapsulate the procedure in a static class.

**Comment by sysexcontrol** [comment]
Yes, this would solve my point.
But I propose to do that in another step, so I will approve this one now.

**Pull request merged**